File: web/app/code-examples/python/connect-agent.py
import asyncio
import json
import sys

import websockets

url = '<WS_URL>'
agent_id = '___PLACEHOLDER_AGENT_ID___'
session_id = ''
user_message = 'Hello!'

chunk_buffers = {}


def handle_streaming_event(event):
    event_type = event.get('event_type')
    if event_type == 'content_block_start':
        pass
    elif event_type == 'content_block_delta':
        text = event.get('delta', {}).get('text')
        if text:
            sys.stdout.write(text)
            sys.stdout.flush()
    elif event_type == 'content_block_stop':
        pass


async def main():
    async with websockets.connect(url) as ws:
        await ws.send(
            json.dumps(
                {
                    'action': 'chat',
                    'agent_id': agent_id,
                    'session_id': session_id,
                    'message': {
                        'message_type': 'text',
                        'content': [{'content_block_type': 'text', 'text': user_message}],
                    },
                }
            )
        )

        async for data in ws:
            msg = json.loads(data)

            if msg.get('chunked'):
                chunk_id = msg['chunk_id']
                chunk_index = msg['chunk_index']
                chunk_total = msg['chunk_total']
                if chunk_id not in chunk_buffers:
                    chunk_buffers[chunk_id] = [None] * chunk_total
                chunk_buffers[chunk_id][chunk_index] = msg['payload']
                if any(p is None for p in chunk_buffers[chunk_id]):
                    continue
                msg = json.loads(''.join(chunk_buffers.pop(chunk_id)))

            if msg['response_type'] == 'data':
                handle_streaming_event(msg['event'])


asyncio.run(main())

File: web/app/code-examples/node/connect-agent.js
import WebSocket from 'ws';

const url = '<WS_URL>';
const agentId = '___PLACEHOLDER_AGENT_ID___';
const sessionId = '';
const userMessage = 'Hello!';

const ws = new WebSocket(url);

ws.addEventListener('open', () => {
  ws.send(
    JSON.stringify({
      action: 'chat',
      agent_id: agentId,
      session_id: sessionId,
      message: {
        message_type: 'text',
        content: [{ content_block_type: 'text', text: userMessage }],
      },
    }),
  );
});

const chunkBuffers = {};

ws.addEventListener('message', (event) => {
  let msg = JSON.parse(event.data);

  if (msg.chunked) {
    const { chunk_id, chunk_index, chunk_total, payload } = msg;
    if (!chunkBuffers[chunk_id]) {
      chunkBuffers[chunk_id] = new Array(chunk_total).fill(null);
    }
    chunkBuffers[chunk_id][chunk_index] = payload;
    if (chunkBuffers[chunk_id].some(p => p === null)) return;
    msg = JSON.parse(chunkBuffers[chunk_id].join(''));
    delete chunkBuffers[chunk_id];
  }

  switch (msg.response_type) {
    case 'data':
      handleStreamingEvent(msg.event);
      break;
  }
});

const handleStreamingEvent = (event) => {
  switch (event.event_type) {
    case 'content_block_start':
      break;
    case 'content_block_delta':
      if (event.delta?.text) {
        process.stdout.write(event.delta.text);
      }
      break;
    case 'content_block_stop':
      break;
  }
};
